sidecar/proxy/monitor/discovery.go

```diff
@@ -142,10 +157,16 @@ func (m *discoveryMonitor) poll() error {
 	m.cache = catalog
 
 	// Notify the listeners
+	var listeners []DiscoveryListener
+	func() {
+		m.mutex.Lock()
+		defer m.mutex.Unlock()
+		listeners = make([]DiscoveryListener, len(m.listeners))
```

Comment:
It is not needed to allocate `listeners` and copy `m.listeners` into it. It is sufficient (and correct) to just copy the `m.listeners` _reference_ into `listeners`. Using copy-on-write during add/remove guarantees that the slice referenced by `listeners` is immutable, so can be iterated-on without a lock held. However, it's still important to copy the slice reference with a lock held, both for atomicity and [happens-before](https://golang.org/ref/mem#tmp_2) relation.